gtk4/src/subclass/widget.rs

```diff
@@ -123,3 +126,127 @@ unsafe extern "C" fn widget_direction_changed<T: ObjectSubclass>(
+        klass: &mut subclass::simple::ClassStruct<Self>,
+        resource_name: &str,
+    ) {
+        unsafe {
+            let type_class = klass as *mut _ as *mut gobject_sys::GTypeClass;
+            let widget_class =
+                gobject_sys::g_type_check_class_cast(type_class, gtk_sys::gtk_widget_get_type())
+                    as *mut gtk_sys::GtkWidgetClass;
+            gtk_sys::gtk_widget_class_set_template_from_resource(
+                widget_class,
+                resource_name.to_glib_none().0,
+            );
+        }
+    }
+
+    fn bind_template_child(klass: &mut subclass::simple::ClassStruct<Self>, name: &str) {
+        unsafe {
+            let type_class = klass as *mut _ as *mut gobject_sys::GTypeClass;
+            let widget_class =
+                gobject_sys::g_type_check_class_cast(type_class, gtk_sys::gtk_widget_get_type())
+                    as *mut gtk_sys::GtkWidgetClass;
+            gtk_sys::gtk_widget_class_bind_template_child_full(
+                widget_class,
+                name.to_glib_none().0,
+                false as glib_sys::gboolean,
+                0,
+            );
+        }
+    }
+
+    unsafe fn bind_template_child_with_offset<T>(
+        klass: &mut subclass::simple::ClassStruct<Self>,
+        name: &str,
+        offset: field_offset::FieldOffset<Self, TemplateChild<T>>,
+    ) where
+        T: ObjectType + FromGlibPtrNone<*mut <T as ObjectType>::GlibType>,
+    {
+        let type_class = klass as *mut _ as *mut gobject_sys::GTypeClass;
+        let widget_class =
+            gobject_sys::g_type_check_class_cast(type_class, gtk_sys::gtk_widget_get_type())
+                as *mut gtk_sys::GtkWidgetClass;
+        let private_offset = Self::type_data().as_ref().private_offset;
+        gtk_sys::gtk_widget_class_bind_template_child_full(
+            widget_class,
+            name.to_glib_none().0,
+            false as glib_sys::gboolean,
+            private_offset + (offset.get_byte_offset() as isize),
+        )
+    }
+}
+
+unsafe impl<T: ObjectSubclass + WidgetImpl> WidgetClassSubclassExt for T {}
+
+#[derive(Debug, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct TemplateChild<T>
+where
+    T: ObjectType + FromGlibPtrNone<*mut <T as ObjectType>::GlibType>,
+{
+    ptr: *mut <T as ObjectType>::GlibType,
+}
+
+impl<T> Default for TemplateChild<T>
+where
+    T: ObjectType + FromGlibPtrNone<*mut <T as ObjectType>::GlibType>,
+{
+    fn default() -> Self {
+        Self {
+            ptr: std::ptr::null_mut(),
+        }
+    }
+}
+
+impl<T> TemplateChild<T>
+where
+    T: ObjectType + FromGlibPtrNone<*mut <T as ObjectType>::GlibType>,
+{
+    #[track_caller]
+    pub fn get(&self) -> T {
+        unsafe {
+            Option::<T>::from_glib_none(self.ptr)
+                .expect("Failed to retrieve template child. Please check that it has been bound.")
+        }
+    }
+}
+
+pub trait CompositeTemplate
+where
+    Self: WidgetClassSubclassExt,
```

Comment:
Why not
```rust
pub trait CompositeTemplate: WidgetClassSubclassExt { ... }
```
?

Also that can just be `WidgetImpl` instead of the extension trait. Equivalent but probably slightly easier to resolve for the compiler.

Reply:
Can it? bind_template_children() in the macro relies on the functions from WidgetClassSubclassExt, so these bounds ensure that it's implemented.

Reply:
That should be something the compiler should be able to figure out, but I've to see it in practice :)